pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/ReferralSourceService.java

```diff
@@ -0,0 +1,408 @@
+                    new ParticipantData.Builder()
+                            .withParticipantDataId(participantData.getParticipantDataId())
+                            .withDdpParticipantId(ddpParticipantId)
+                            .withDdpInstanceId(participantData.getDdpInstanceId())
+                            .withFieldTypeId(RGP_PARTICIPANT_DATA)
+                            .withData(gson.toJson(props))
+                            .withLastChanged(System.currentTimeMillis())
+                            .withChangedBy(userId).build());
+        } catch (JsonSyntaxException | JsonIOException | ClassCastException e) {
+            throw new DsmInternalError("Invalid data format " + msg, e);
+        }
+        return true;
+    }
+
+    protected static Map<String, String> getDataMap(ParticipantData participantData) throws JsonSyntaxException {
+        Optional<String> data = participantData.getData();
+        if (data.isEmpty() || StringUtils.isEmpty(data.get())) {
+            return Collections.emptyMap();
+        }
+
+        return gson.fromJson(data.get(), Map.class);
+    }
+
+    private List<Activities> getParticipantActivities(String ddpParticipantId, String esIndex) {
+        Optional<ElasticSearchParticipantDto> esParticipant =
+                ElasticSearchUtil.getParticipantESDataByParticipantId(esIndex, ddpParticipantId);
+        if (esParticipant.isEmpty()) {
+            throw new ESMissingParticipantDataException("Participant ES data is null for participant " + ddpParticipantId);
+        }
+        return esParticipant.get().getActivities();
+    }
+
+    /**
+     * Derive REFERRAL_SOURCE ID from participant activities
+     *
+     * @param activities list of DSS participant activities
+     * @return a referral source ID
+     * @throws DsmInternalError If the expected referral source values are not present or the referral source mapping is
+     *                          missing or out of sync
+     */
+    public static String deriveReferralSourceId(List<Activities> activities) {
+        return ReferralSourceService.convertReferralSources(ReferralSourceService.getReferralSources(activities));
+    }
+
+    /**
+     * Get ref sources from activities. Specifically get answers to FIND_OUT enrollment question.
+     *
+     * @return list of strings, empty if question or answers not found
+     */
+    protected static List<String> getReferralSources(List<Activities> activities) {
+        Optional<Activities> enrollmentActivity = activities.stream().filter(activity ->
+                DDPActivityConstants.ACTIVITY_ENROLLMENT.equals(activity.getActivityCode())).findFirst();
+        if (enrollmentActivity.isEmpty()) {
+            log.warn("Could not derive referral source data, participant has no enrollment activity");
+            return new ArrayList<>();
+        }
+        List<Map<String, Object>> questionsAnswers = enrollmentActivity.get().getQuestionsAnswers();
```

Comment:
I only see one other use of a pattern like this in `ClinicalKitDto`.  I can refactor to have both use the same code, but since `ClinicalKitDto` lacks automated testing, I am reluctant to make arbitrary changes.

Reply:
sounds good. we can leave as is, no refactor needed